### DISCUSSION TITLE
Gradle rules

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#5494] [General]
Gradle has a lot of foot gun apis. An example: `TaskContainer.create` instead of `TaskContainer.register`. And a lot of special ways to declare things. 

I think that detekt should be able to help to write better gradle code (plugins and scripts). But I'm not sure how. Probably with a configuration of forbidden method call a lot of use cases would be solved. But it's difficult to share configurations. Other option is a first/third party plugin with a copy of forbidden method call and its own default values. Easier to share.

But we would still have another issue. Our gradle plugin doesn't have a task to check the gradle scripts. That's a problem because Forbidden method call needs type solving. The work around would be to use the cli. But it's difficult to configure.

Should we create issues to track the different stoppers? Does it have sense to use detekt for this? Are there other options out there? (I did a little research and found nothing)

**Comment by schalkms** [comment]
In my opinion, it makes sense to include a **Gradle** ruleset in detekt.

**Comment by cortinico** [comment]
That's a great idea. Perhaps we can get inspired by some of rules from: 
https://github.com/runningcode/gradle-doctor

**Comment by 3flex** [comment]
It's a good idea, but before duplicating effort, take a look at https://github.com/autonomousapps/gradle-best-practices-plugin. It already has some checks for `create` instead of `register`, plus others.

The author of this plugin also wrote the popular dependency analysis plugin (https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin).

**Reply by 3flex** [reply]
This doesn't check Gradle scripts though AFAIK, just Gradle plugins. But I don't think setting detekt up to check Gradle scripts with type resolution will be easy.
